ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs

```diff
@@ -0,0 +1,194 @@
+
+import           Ouroboros.Network.Chain (Chain (..))
+
+import           Ouroboros.Storage.ChainDB.API (ChainDB)
+import qualified Ouroboros.Storage.ChainDB.API as ChainDB
+import qualified Ouroboros.Storage.ChainDB.Mock as Mock
+
+import           Test.Util.TestBlock (GenTx (..), GenTxId (..), TestBlock,
+                     computeGenTxId, singleNodeTestConfig, testInitExtLedger)
+import           Test.Util.TestTx (TestTx (..))
+
+tests :: TestTree
+tests = testGroup "Mempool"
+  [ testProperty "getTxs == getTxsAfter zeroIdx"        prop_Mempool_getTxs_getTxsAfter
+  , testProperty "addedTxs == getTxs"                   prop_Mempool_addTxs_getTxs
+  , testProperty "Invalid transactions are never added" prop_Mempool_InvalidTxsNeverAdded
+  ]
+
+{-------------------------------------------------------------------------------
+  Mempool Implementation Properties
+-------------------------------------------------------------------------------}
+
+-- | Test that @getTxs == getTxsAfter zeroIdx@.
+prop_Mempool_getTxs_getTxsAfter :: Chain TestBlock
+                                -> [TestTx]
+                                -> Property
+prop_Mempool_getTxs_getTxsAfter bc txs =
+  testAddTxsWithMempoolAndSnapshot
+    bc
+    txs
+    (\Mempool{zeroIdx} MempoolSnapshot{getTxs, getTxsAfter} ->
+      getTxs === getTxsAfter zeroIdx)
+
+-- | Test that all valid transactions added to a 'Mempool' can be retrieved
+-- afterward.
+--
+-- n.b. In this test, we add a list of arbitrary transactions to an initially
+-- empty 'Mempool', retrieve all of the transactions from the 'Mempool', and
+-- assert that those transactions are /specifically/ equal to the /valid/
+-- transactions that we attempted to add (the invalid ones should have been
+-- dropped).
+prop_Mempool_addTxs_getTxs :: Chain TestBlock
+                           -> [TestTx]
+                           -> Property
+prop_Mempool_addTxs_getTxs bc txs =
+  testAddTxsWithMempoolAndSnapshot
+    bc
+    txs
+    (\_ MempoolSnapshot{getTxs} ->
+      filter (genTxIsValid . snd) (testTxsToGenTxPairs txs)
+        === dropThd getTxs)
+ where
+  dropThd :: [(a, b, c)] -> [(a, b)]
+  dropThd = map (\(a, b, _) -> (a, b))
+  --
+  genTxIsValid :: GenTx TestBlock -> Bool
+  genTxIsValid (TestGenTx (ValidTestTx _)) = True
+  genTxIsValid _                           = False
+
+-- | Test that invalid transactions can never be added to the 'Mempool'.
+--
+-- n.b. In this test, we add an arbitrary list of transactions to an initially
+-- empty 'Mempool', retrieve all of the transactions from the 'Mempool', and
+-- assert that none of those transactions are invalid.
+prop_Mempool_InvalidTxsNeverAdded :: Chain TestBlock
+                                  -> [TestTx]
+                                  -> Property
+prop_Mempool_InvalidTxsNeverAdded bc txs =
+  testAddTxsWithMempoolAndSnapshot
+    bc
+    txs
+    (\_ MempoolSnapshot{getTxs} ->
+      any (\(_, tx, _) -> genTxIsInvalid tx) getTxs === False)
```

Comment:
Tip: if this test fails, what will be printed? Something like `True /== False`, which doesn't give you any useful information. If you change the condition to `filter (\(_, tx, _) -> genTxIsInvalid tx) getTxs === []`, you'll get: `[<some invalid transaction>, ..] /== []`, which is much more useful.

While writing the above, I realised that you already print the invalid transactions in `counterexampleMempoolInfo` 👍. So you can leave this as is 🙂

Reply:
Either way I think your suggestion is nicer. Will go that route. Thanks.